Kernel/VirtIO/VirtIOConsole.cpp

```diff
@@ -73,115 +60,135 @@ bool VirtIOConsole::handle_device_config_change()
+    } else if (queue_index == CONTROL_TRANSMITQ) {
+        ScopedSpinLock ringbuffer_lock(m_control_transmit_buffer->lock());
+        auto& queue = get_queue(CONTROL_TRANSMITQ);
         ScopedSpinLock queue_lock(queue.lock());
         size_t used;
         VirtIOQueueChain popped_chain = queue.pop_used_buffer_chain(used);
         do {
             popped_chain.for_each([this](PhysicalAddress address, size_t length) {
-                m_transmit_buffer->reclaim_space(address, length);
+                m_control_transmit_buffer->reclaim_space(address, length);
             });
             popped_chain.release_buffer_slots_to_queue();
             popped_chain = queue.pop_used_buffer_chain(used);
         } while (!popped_chain.is_empty());
-        // Unblock any IO tasks that were blocked because can_write() returned false
-        evaluate_block_conditions();
-        break;
+    } else {
+        auto port_index = queue_index / 2 == 0 ? 0 : (queue_index - 2) / 2;
+        m_ports.get(port_index).value()->handle_queue_update({}, queue_index);
     }
-    default:
-        VERIFY_NOT_REACHED();
-    }
-}
-
-bool VirtIOConsole::can_read(const FileDescription&, size_t) const
-{
-    return m_receive_buffer->used_bytes() > 0;
 }
 
-KResultOr<size_t> VirtIOConsole::read(FileDescription& desc, u64, UserOrKernelBuffer& buffer, size_t size)
+void VirtIOConsole::setup_multiport()
 {
-    if (!size)
-        return 0;
+    m_control_receive_buffer = make<RingBuffer>("VirtIOConsole control receive queue", CONTROL_BUFFER_SIZE);
+    m_control_transmit_buffer = make<RingBuffer>("VirtIOConsole control receive queue", CONTROL_BUFFER_SIZE);
 
-    if (!can_read(desc, size))
-        return EAGAIN;
-
-    ScopedSpinLock ringbuffer_lock(m_receive_buffer->lock());
-    auto& queue = get_queue(RECEIVEQ);
+    auto& queue = get_queue(CONTROL_RECEIVEQ);
     ScopedSpinLock queue_lock(queue.lock());
+    VirtIOQueueChain chain(queue);
+    auto offset = 0ul;
+    auto buffer_size = 8;
 
-    auto bytes_copied = m_receive_buffer->copy_data_out(size, buffer);
-    m_receive_buffer->reclaim_space(m_receive_buffer->start_of_used(), bytes_copied.value());
-    evaluate_block_conditions();
+    while (offset < CONTROL_BUFFER_SIZE) {
+        auto buffer_start = m_control_receive_buffer->start_of_region().offset(offset);
+        auto did_add_buffer = chain.add_buffer_to_chain(buffer_start, buffer_size, BufferType::DeviceWritable);
+        VERIFY(did_add_buffer);
+        offset += buffer_size;
+        supply_chain_and_notify(CONTROL_RECEIVEQ, chain);
+    }
 
-    return bytes_copied;
+    ControlMessage ready_event {
+        .id = 0, /* unused */
+        .event = (u16)ControlEvent::DeviceReady,
+        .value = 1 /* success */
+    };
+    write_control_message(ready_event);
 }
-
-bool VirtIOConsole::can_write(const FileDescription&, size_t) const
+void VirtIOConsole::process_control_message(VirtIOConsole::ControlMessage message)
```

Comment:
no need to explicitly specify `VirtIOConsole::`

Reply:
Also, minor nitpick of no empty line between the methods.